magda-io/magda
# Issue/2399 Add Access Location AutoComplete input to Dataset Page 4 (Access and User)

### What this PR does

Ready to review & merge now~

Fixes #2399 

Add Access Location AutoComplete input to Dataset Page 4 (Access and User)
- User Edge Ngram token filter to implement auto-complete API
  - Why not [completion suggester](https://www.elastic.co/guide/en/elasticsearch/reference/6.4/search-suggesters-completion.html)
    - It does not inherit search/query context and there is no easy way to implement access filters (Es team believe it will defeat the purpose of having a separate `suggester` interface). 
    - `Context Suggester` is a different filter interface not for complex query
  - Why not [Match phrase prefix query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query-phrase-prefix.html)?
    - It's a search time feature. Thus, much slower than other options
  - Why use [Edge Ngram token filter](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-edgengram-tokenizer.html)?
    - It's an index time feature. Thus, run much faster at searching time
    - As it's indexing feature, it doesn't impact required filtering logic. i.e. the same access control queries can be used.

Based on @alyce 's new design:

![IMG_3468](https://user-images.githubusercontent.com/40975814/63076026-0a84ce80-bf77-11e9-8f08-202eca16bdbd.jpg)

### Test Notes:

- Make sure you are logged in as `admin` users. Otherwise, you won't be able to access `add dataset` page (You probably will see a loading screen see issue #2406 ).
- Make sure your account has the `approver` role --- otherwise, you can access `draft datasets`. Thus, no suggestion items will be generated from the draft datasets that you just created 

### Test Site:

**To be available...**

### Checklist
